### PR TITLE
[Music Lab] Add setEffect block

### PR DESCRIPTION
**Pull request by mikeharv**
@onlinecsteacher has request effect blocks in the advanced mode toolbox. The new block presented here looks just like its Simple2 counterpart, with some slight difference when it comes to the function of the command called by its generated code.

In Simple2, functions managed an effect stack. If an effect is set within a function, it can only be applied to blocks from that same scope. To facilitate this, `setEffect` pushes to a stack. When a sound is played, the effects item in the stack is utilized. When a function begins or ends, the current effect is pushed or popped.

In Advanced, we let the student handle the scope of their effect changes. As such, we only need to track one set of effects.

image

## Links

https://docs.google.com/document/d/1NROqFS9gA4wklBNsxdrFQbou9HHI-fwPEetKGNm-Eho/edit

## Testing story

Tested a variety of programs.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/music/blockly/blockUtils.js [hunk 1]
```diff
@@ -50,16 +50,9 @@ export function getCodeForSingleBlock(block) {
...
-      return true;
-    }
-  }
-
-  return false;
+// Check if root block is 'when_run'.
+export const isBlockInsideWhenRun = block => {
+  return [BlockTypes.WHEN_RUN, BlockTypes.WHEN_RUN_SIMPLE2].includes(
+    block.getRootBlock().type
+  );
```
Comment (mikeharv) [inline apps/src/music/blockly/blockUtils.js:57]: This change is unrelated to the PR. Instead of recursively checking the next parent block, we can simplify by checking the root (top-most) block directly.

File: apps/src/music/blockly/blocks/advanced.js [hunk 2]
```diff
@@ -42,8 +45,6 @@ export const playSound = {
       'measure',
       Blockly.JavaScript.ORDER_ASSIGNMENT
     ) +
-    ', ' +
-    (isBlockInsideWhenRun(ctx) ? 'true' : 'false') +
```
Comment (mikeharv) [inline apps/src/music/blockly/blocks/advanced.js:46]: This change is unrelated to the PR. This argument is not actually used in `playSoundAtMeasureById`.

File: apps/src/music/player/sequencer/AdvancedSequencer.ts [hunk 3]
```diff
@@ -13,20 +14,17 @@ import Sequencer from './Sequencer';
...
+    this.effects = {};
   }
 
-  playSoundAtMeasureById(
-    id: string,
-    measure: number,
-    isBlockInsideWhenRun: boolean,
-    blockId: string
-  ) {
+  playSoundAtMeasureById(id: string, measure: number, blockId: string) {
```
Comment (mikeharv) [inline apps/src/music/player/sequencer/AdvancedSequencer.ts:27]: This change is unrelated to the PR. The `isBlockInsideWhenRun` parameter is not actually used in this function.

**Comment by sanchitmalhotra126** [comment]
> In Advanced, we let the student handle the scope of their effect changes. As such, we only need to track one set of effects.

This is interesting. This effectively means that all effect commands are global right? And an effect set in function would overwrite effects set under when_run (or vice versa depending on call order)?

**Comment by mikeharv** [comment]
@sanchitmalhotra126 That’s correct. @breville and I discussed on a call how Simple2 is designed to handle some of the magic for students automatically. In this model, you can achieve the same thing by manually resetting the effect at the end of the context stack (function, trigger, when run) that originally sets it.

File: apps/src/music/blockly/blockUtils.js (hunk 1 above)
Comment (sanchitmalhotra126) [reply apps/src/music/blockly/blockUtils.js:57]: Neat!

**Comment by breville** [comment]
Can we ensure that the current effects are reset to defaults at the beginning of each handler (for `when run` or a trigger)?  From a quick test it appears that otherwise we'll be using the last set effect, even on a fresh run of the program.

**Comment by mikeharv** [comment]
> Can we ensure that the current effects are reset to defaults at the beginning of each handler (for `when run` or a trigger)? From a quick test it appears that otherwise we'll be using the last set effect, even on a fresh run of the program.

Thanks for testing, and good catch! It looks resetting the effects during `clear()` is enough to take care of this. I've confirmed that now the effects don't carry over between handlers, including on a fresh run of the program.

**Review by breville — APPROVED** [review]
